refactor: aztec l1 contracts deployed w/ forge

ludamad opened the pull request:
Part of https://linear.app/aztec-labs/issue/A-316/consolidate-contract-deployments

This PR refactors the deployment pipeline for governance and rollup contracts. The typescript deployment mechanism is fully replaced with Forge. The forge scripts take configuration using environment variables. They write their deployed contract addresses to a JSON file, parsed by the caller. The results are intended to be exactly the same, barring minor things (e.g. deploying the mock zkpassport verifier in more cases). 

Core files added:
    - DeployAztecL1Contracts.s.sol - Full deployment (fresh network)
    - DeployRollupForUpgrade.s.sol - Upgrade-only deployment
    - DeployRollupLib.sol - Shared rollup deployment library
    - DeploymentConfiguration.sol - Full deployment config. Reads env vars, provides defaults based on network label
    - RollupConfiguration.sol - Rollup-specific config (genesis state, staking stuff, initial validators etc.)

Then lots of work making sure the right glue was in place for the typescript side.

#
Why move to Forge for deployments?

L1 contract deployment was being done in TypeScript but clearly from the network launching using foundry scripts, this was not ideal. 

Benefits:
 - forge --verify automatically posts to etherscan for us, avoiding the complicated custom logic we had to implement
 - forge simulation means we catch any problems in the entire deployment flow early. Stuff like the previous validateConfig() that guarded against reverts are now redundant.
 - forge is able to broadcast these in transaction batches, I got 2-3 slots on Sepolia for a deploy (24-40s)
 - better separation of concerns, l1-contracts now can fully own the deployment
 - foundry tests are very effective at asserting things about our deployments under different conditions

Weirdness:
 - we work around a problem in solidity where MockBlobLib was not deployable through forge but it feels a bit hacky. I think it's preferable to needing to rebuild with a different forge profile like ignition-monorepo did.

Cut features:
- Deploying with a salt is no longer supported. While nice for deterministic deployments, the reason the salt was useful before was mainly to batch deployments. We do this with Forge, so forego the complexity of supporting this. 
This is feasible to bring back if it comes up, just would add a bunch of conditionality to the script.

Comments on future work:
 - It feels like the definition of networks is a bit sprawling and a bit hard to analyze, as the env vars that trigger deployment can be set very high up in the stack, in e.g. test scenario YAMLs. It would be best IMO to make each value only defined in a specific module. Given our limited set of distinct scenarios, we could lean in to the network label approach like in previous ethereum/config.ts (and now DeploymentConfiguration).

In that world, l1-contracts would choose the parameters based on a network label and no other part of the stack could inject values thru env vars, making analysis easier. This would be more like the ignition-monorepo deploys (but they had fewer variations to contend with).

For now it is a bit of a mix as network-specific governance configuration definitions (local, staging, testnet, mainnet)  from TypeScript's config.ts have been moved to DeploymentConfiguration.sol but it allows for a lot to be overridden with env variables.

ludamad commented on l1-contracts/src/core/Rollup.sol:
```diff
@@ -18,7 +18,9 @@ import {IStaking, AttesterConfig, Exit, AttesterView, Status} from "@aztec/core/
 import {IValidatorSelection, IEmperor} from "@aztec/core/interfaces/IValidatorSelection.sol";
 import {IVerifier} from "@aztec/core/interfaces/IVerifier.sol";
 import {TempCheckpointLog, CheckpointLog} from "@aztec/core/libraries/compressed-data/CheckpointLog.sol";
```
I wasn't able to get forge scripts to work properly with the blanket FeeLib.sol, but I am probably missing an obvious setup step

ludamad commented:
<!-- CI3_FLAKE_DETECTION_COMMENT -->
## Flakey Tests

🤖 says: This CI run detected **2** tests that failed, but were tolerated due to a .test_patterns.yml entry.

```
\033FLAKED\033 (\0338;;http://ci.aztec-labs.com/0afb159fd916ea900afb159fd916ea908;;\033):  yarn-project/end-to-end/scripts/run_test.sh simple src/e2e_epochs/epochs_proof_fails.test.ts (603s) (code: 124) group:e2e-p2p-epoch-flakes (\033ludamad\033: fmt)
\033FLAKED\033 (\0338;;http://ci.aztec-labs.com/9ffd07a96179abe49ffd07a96179abe48;;\033):  yarn-project/end-to-end/scripts/run_test.sh simple src/e2e_epochs/epochs_multiple.test.ts (605s) (code: 124) group:e2e-p2p-epoch-flakes (\033ludamad\033: fmt)
```

ludamad commented on yarn-project/ethereum/src/deploy_l1_contract.ts:
```diff
@@ -0,0 +1,362 @@
+import { EthAddress } from '@aztec/foundation/eth-address';
+import { type Logger, createLogger } from '@aztec/foundation/log';
+import { DateProvider } from '@aztec/foundation/timer';
+
```
This is moved code, from deploy_l1_contracts.ts

alexghr commented on l1-contracts/src/mock/libraries/BlobLib.sol:
```diff
@@ -45,8 +44,20 @@ library BlobLib {
+        return abi.decode(data, (uint256));
+      }
+      // During broadcast to anvil, the cheatcode returns empty data. Return a sensible default.
+      return 1;
```
This is probably not needed for the eth-devnets we deploy (e.g. in the tps-scenario network) because we use reth/lighthouse there and blobs should be fully supported.

ludamad replied:
from DMs:
```
this has to be built in for tests to work but it has to be deep in bloblib but if its in then deployments dont work because forge always believes this address exists during simulations
the alternative is to use FORGE_PROFILE=production but that requires that ugly force build hack
so this is needed to be able to deploy the tested code without rebuild to the scenario networks
if we really wanted we can deploy production and avoid those on our scenario networks but you end up having to rebuild l1 contracts entirely
```

alexghr commented on yarn-project/end-to-end/out:
👀

alexghr commented on yarn-project/ethereum/src/deploy_aztec_l1_contracts.ts:
```diff
@@ -0,0 +1,617 @@
+
+// Covers an edge where where we may have a cached BlobLib that is not meant for production.
+// Despite the profile apparently sometimes cached code remains (so says Lasse after his ignition-monorepo arc).
+async function maybeForgeForceForceProductionBuild(l1ContractsPath: string, script: string, chainId: number) {
```
One 'force' too many or intended? 😆

alexghr commented on yarn-project/ethereum/src/deploy_aztec_l1_contracts.ts:
```diff
@@ -0,0 +1,617 @@
+    );
+
+    readline.createInterface({ input: proc.stdout }).on('line', logger.info.bind(logger));
+    readline.createInterface({ input: proc.stderr }).on('line', logger.error.bind(logger));
```
Minor nit: We do this in a bunch of different places in this file. I think a `runForge` function would have helped.

alexghr commented on yarn-project/ethereum/src/deploy_l1_contract.ts:
is this just `_contracts` -> `_contract`? If we don't use this flow anymore I think we should just delete it. We can always restore from git history if needed

ludamad replied:
No that is not it. we still use the single deploy code currently for some one offs

alexghr commented on yarn-project/out:
👀

alexghr reviewed (commented):
Looks good. I think once the `out` files are deleted this PR can be merged.

alexghr commented:
I forgot to mention in the review: with salt removed we have to refactor network deployments because they rely on salt for idempotent contract deployments.

just-mitch commented on l1-contracts/script/deploy/DeployAztecL1Contracts.s.sol:
```diff
@@ -0,0 +1,315 @@
+        output.governance = new Governance(
+            output.stakingAsset,
+            address(output.governanceProposer),
+            address(output.gse),
```
Alert! This needs to be `address(0)`. Setting it as such allows anyone to participate in governance (which is what we want).

ludamad replied:
Fixed, good catch :)

just-mitch commented on spartan/scripts/deploy_network.sh:
```diff
@@ -274,7 +273,7 @@ fi
 ROLLUP_CONTRACTS_START=$(date +%s)
 DEPLOY_ROLLUP_CONTRACTS_DIR="${SCRIPT_DIR}/../terraform/deploy-rollup-contracts"
-"${SCRIPT_DIR}/override_terraform_backend.sh" "${DEPLOY_ROLLUP_CONTRACTS_DIR}" "${CLUSTER}" "${BASE_STATE_PATH}/deploy-rollup-contracts/${SALT}"
+"${SCRIPT_DIR}/override_terraform_backend.sh" "${DEPLOY_ROLLUP_CONTRACTS_DIR}" "${CLUSTER}" "${BASE_STATE_PATH}/deploy-rollup-contracts/$ROLLUP_CONTRACTS_START"
```
This is nice. Did/can you verify that terraform actually never retries if the job succeeds once?

ludamad replied:
Sorry, what do you mean exactly by 'this job', the l1 contracts deploy?

just-mitch replied:
Yes

just-mitch replied:
I mean the `tf_run` command below. In the old flow, it would just rerun the deployment, but since the salts/deployed contracts were all the same, it was basically a no op (but the container still ran deploy l1 contracts). I think without the salt, we should configure at the terraform level (somehow) "once this job succeeds, never try again"

just-mitch replied:
I also think it is okay to do that in a separate PR. It only really matters for the long-standing networks, and in the immediate term we can just deploy the contracts manually, and then set `CREATE_ROLLUP_CONTRACTS=false` in their .env files

ludamad replied:
as mentioned, I've tackled this with a simple read of tfstate before the run

ludamad committed:
refactor: aztec l1 contracts deployed w/ forge

Part of https://linear.app/aztec-labs/issue/A-316/consolidate-contract-deployments

This PR refactors the deployment pipeline for governance and rollup contracts. The typescript deployment mechanism is fully replaced with Forge. The forge scripts take configuration using environment variables. They write their deployed contract addresses to a JSON file, parsed by the caller. The results are intended to be exactly the same, barring minor things (e.g. deploying the mock zkpassport verifier in more cases).

Core files added:
    - DeployAztecL1Contracts.s.sol - Full deployment (fresh network)
    - DeployRollupForUpgrade.s.sol - Upgrade-only deployment
    - DeployRollupLib.sol - Shared rollup deployment library
    - DeploymentConfiguration.sol - Full deployment config. Reads env vars, provides defaults based on network label
    - RollupConfiguration.sol - Rollup-specific config (genesis state, staking stuff, initial validators etc.)

Then lots of work making sure the right glue was in place for the typescript side.

#
Why move to Forge for deployments?

L1 contract deployment was being done in TypeScript but clearly from the network launching using foundry scripts, this was not ideal.

Benefits:
 - forge --verify automatically posts to etherscan for us, avoiding the complicated custom logic we had to implement
 - forge simulation means we catch any problems in the entire deployment flow early. Stuff like the previous validateConfig() that guarded against reverts are now redundant.
 - forge is able to broadcast these in transaction batches, I got 2-3 slots on Sepolia for a deploy (24-40s)
 - better separation of concerns, l1-contracts now can fully own the deployment
 - foundry tests are very effective at asserting things about our deployments under different conditions

Weirdness:
 - we work around a problem in solidity where MockBlobLib was not deployable through forge but it feels a bit hacky. I think it's preferable to needing to rebuild with a different forge profile like ignition-monorepo did.

Cut features:
- Deploying with a salt is no longer supported. While nice for deterministic deployments, the reason the salt was useful before was mainly to batch deployments. We do this with Forge, so forego the complexity of supporting this.
This is feasible to bring back if it comes up, just would add a bunch of conditionality to the script.

Comments on future work:
 - It feels like the definition of networks is a bit sprawling and a bit hard to analyze, as the env vars that trigger deployment can be set very high up in the stack, in e.g. test scenario YAMLs. It would be best IMO to make each value only defined in a specific module. Given our limited set of distinct scenarios, we could lean in to the network label approach like in previous ethereum/config.ts (and now DeploymentConfiguration).

In that world, l1-contracts would choose the parameters based on a network label and no other part of the stack could inject values thru env vars, making analysis easier. This would be more like the ignition-monorepo deploys (but they had fewer variations to contend with).

For now it is a bit of a mix as network-specific governance configuration definitions (local, staging, testnet, mainnet)  from TypeScript's config.ts have been moved to DeploymentConfiguration.sol but it allows for a lot to be overridden with env variables.
